src/student_success_tool/modeling/bias_detection.py

```diff
@@ -0,0 +1,213 @@
+        cm.ravel()
+        if cm.shape == (2, 2)
+        else np.bincount(np.array(targets) * 2 + np.array(preds), minlength=4)
+    )
+
+    denominator = fn + tp
+
+    if (tp < min_fnpr_samples) or (fn < min_fnpr_samples):
+        return np.nan, np.nan, np.nan
+
+    fnpr = fn / denominator
+    margin = Z * np.sqrt((fnpr * (1 - fnpr)) / denominator)
+    return fnpr, max(0, fnpr - margin), min(1, fnpr + margin)
+
+
+def check_ci_overlap(
+    ci1: tuple[float, float],
+    ci2: tuple[float, float],
+) -> bool:
+    """
+    Checks whether confidence intervals (CIs) overlap. If they do, the FNPR differences
+    are within the margin of error at the 95% confidence level. If the CIs do not
+    overlap, this suggests strong statistical evidence that the FNPRs are different.
+
+    Args:
+        ci1: Confidence interval (min, max) for subgroup 1
+        ci2: Confidence interval (min, max) for subgroup 2
+    Returns:
+        Boolean indicating whether the CIs overlap.
+    """
+    return not (ci1[1] < ci2[0] or ci2[1] < ci1[0])
+
+
+def z_test_fnpr_difference(
+    fnpr1: float,
+    fnpr2: float,
+    denominator1: int,
+    denominator2: int,
+) -> float:
+    """
+    Performs a z-test for the FNPR difference between two groups. If there are
+    less than 30 samples of false negatives and true negatives, then we do not
+    have enough data to perform a z-test. Thirty samples is the standard check
+    for z-tests.
+
+    Args:
+        fnpr1: FNPR value for subgroup 1
+        fnpr2: FNPR value for subgroup 2
+        denominator1: Number of false negatives + true negatives for subgroup 1
+        denominator2: Number of false negatives + true negatives for subgroup 2
+    Returns:
+        Two-tailed p-value for the z-test for the FNPR difference between the two subgroups.
+    """
+    if (
+        denominator1 <= 30 or denominator2 <= 30
+    ):  # Ensures valid sample sizes for z-test
+        return np.nan
+    std_error = np.sqrt(
+        (fnpr1 * (1 - fnpr1)) / denominator1 + (fnpr2 * (1 - fnpr2)) / denominator2
+    )
+    z_stat = (fnpr1 - fnpr2) / std_error
+    return float(2 * (1 - st.norm.cdf(abs(z_stat))))  # Two-tailed p-value
+
+
+def log_bias_flag(
+    group: str,
+    subgroup1: str,
+    subgroup2: str,
+    fnpr_diff: float,
+    bias_type: str,
+    dataset: str,
+    flag: str,
+    p_value: float = np.nan,
+) -> dict:
+    """
+    Aggregate bias flag information for a given subgroup pair into a dict.
+
+    Args:
+        group: Name of the group (e.g. "Gender", "Race", "Age")
+        subgroup1: Name of the subgroup 1 (e.g. "Female", "Male", "Asian", "African American", "Caucasian")
+        subgroup2: Name of the subgroup 2 (e.g. "Female", "Male", "Asian", "African American", "Caucasian")
+        fnpr_diff: Absolute value of difference in FNPR
+        bias_type: Type of bias (e.g. "Non-overlapping CIs", "Overlapping: : p-value: ...")
+        dataset: Name of the dataset (e.g. train/test/validate)
```

Comment:
this could be confusing, since datasets themselves are named, but here you seem to be talking about "splits"